scipy/signal/spectral.py

```diff
@@ -281,19 +452,45 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
                       % (nperseg, axis, x.shape[axis], axis))
         nperseg = x.shape[-1]
 
+    # These checks only neccesary if x!=y
+    if not psd:
+        if y.shape[-1] < nperseg:
+            warnings.warn('nperseg = %d, is greater than y.shape[%d] = %d, '
+                          'using nperseg = y.shape[%d]'
+                          % (nperseg, axis, y.shape[axis], axis))
+            nperseg = y.shape[-1]
+
+        # Check if we can broadcast the remaining axes together
+        for a, b in zip(x.shape[-2::-1],y.shape[-2::-1]):
+            if a == 1 or b == 1 or a == b:
+                pass
+            else:
+                raise ValueError('x and y cannot be broadcast together.')
+
+        # Check if x and y are the same length, pad if neccesary
+        if x.shape[-1] != y.shape[-1]:
+            if x.shape[-1] < y.shape[-1]:
+                padShape = x.shape
+                padShape[-1] = y.shape[-1] - x.shape[-1]
+                x = np.concatenate((x, np.zeros(padShape)), -1)
+            else:
+                padShape = y.shape
+                padShape[-1] = x.shape[-1] - y.shape[-1]
+                y = np.concatenate((y, np.zeros(padShape)), -1)
+
     if isinstance(window, string_types) or type(window) is tuple:
         win = get_window(window, nperseg)
     else:
         win = np.asarray(window)
         if len(win.shape) != 1:
-            raise ValueError('window must be 1-D')
+            raise ValueError('Window must be 1-D')
         if win.shape[0] > x.shape[-1]:
-            raise ValueError('window is longer than x.')
+            raise ValueError('Window is longer than input data.')
         nperseg = win.shape[0]
 
-    # numpy 1.5.1 doesn't have result_type.
-    outdtype = (np.array([x[0]]) * np.array([1], 'f')).dtype.char.lower()
-    if win.dtype != outdtype:
+    outdtype = np.result_type(np.array([x[0]*y[0]]) * np.array([1], 'F'))
```

Comment:
This should be something more like: `np.result_type(x, y, np.complex64)`  Here you're passing only a single array to result type, and actually calculating the dtype yourself.